instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/snippet/SnippetInjectingPrintWriter.java

```diff
@@ -0,0 +1,46 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+package io.opentelemetry.javaagent.instrumentation.servlet.v3_0.snippet;
+
+import java.io.PrintWriter;
+
+public class SnippetInjectingPrintWriter extends PrintWriter {
+  private final String snippet;
+  private InjectionState state = null;
+
+  public SnippetInjectingPrintWriter(
+      PrintWriter writer,
+      String snippet,
+      String characterEncoding,
+      SnippetInjectingResponseWrapper wrapper) {
+    super(writer);
+    state = new InjectionState(characterEncoding);
+    this.snippet = snippet;
+  }
+
+  @Override
+  public void write(String s, int off, int len) {
+    for (int i = off; i < s.length() && i - off < len; i++) {
+      char b = s.charAt(i);
+      write(b);
+    }
+  }
+
+  @Override
+  public void write(int b) {
+    boolean shouldInject = state.processByte((byte) b);
+    super.write(b);
+    if (shouldInject) {
+      // begin to insert
+      state.setAlreadyInjected(); // set before write to avoid recursive loop
```

Comment:
```suggestion
      // set before write to avoid recursive loop since super.write(String) may delegate back to write(int)
      state.setAlreadyInjected();
```